new provider: dockerregistry

glasser opened the pull request:
(Initial draft. This commit does not include doc updates, and all of the
code is in a single file instead of being split into provider/resource
files as is the terraform standard. It also uses the V1 API rather than
the V2 API mostly because it is simpler to implement, but it might be
better to use the V2 API (eg, Amazon ECR only implements V1).)

Unlike the 'docker' provider, which talks to your local Docker daemon to
read and write your local image stores, the 'dockerregistry' provider
talks to a Docker Registry V1 server and confirms that the named
repository/tag exists.

It is a read-only provider: it doesn't push to the registry; it just
helps you assert that the docker tag push via your build process
actually exists before creating other configuration that uses it.

glasser commented:
I'm submitting this partially because it may be useful to others, but also to try to get feedback on the approach that it takes.  We are using this in our setup in order to validate docker image names (*not managed by terraform*) that are used in the configuration of other resources (EC2 userdata, AWS ECS services, etc).

It works ... OK.  In an ideal world we'd be able to run all of these read-only "validation" checks before making any changes.  And the state stored for these resources doesn't actually represent real state.  Is there a better way to do this sort of read-only validation in Terraform?

(BTW, I originally wrote this against the Docker V2 Registry API using https://github.com/heroku/docker-registry-client instead of the V1 API.  The advantage to using V1 is that you don't need to authenticate for read-only operations on public images, and also the particular library I used requires you to vendor an old version of the docker repo.  But it might make sense to use the V2 version if this is actually accepted into Terraform, especially since AWS ECR only supports V2.)

jen20 commented:
Hi @glasser! It seems that what you want for the "read only" part is #4169? We could use that to support reading docker image data from a repository.

CpuID commented:
A nice idea - I would vote for a V2 only version though (being V1 is effectively obsolete now).

glasser commented:
@jen20 Absolutely, #4169 seems perfect for this sort of thing.

@CpuID I actually originally wrote it for V2 and rewrote to V1 mostly because the library I was using was hard to use due some vendoring issues.  I just forked the library, deleted the parts of it that have the annoying dependencies, and updated this PR to use my fork to implement V2.  I was also incorrect when I stated above that V2 made you authenticate for read-only operations on public repositories (it does make you get a bearer token from their oauth server, but you don't need to authenticate while you do that).

Given that this would be a better fit for #4169 does that mean that this is unlikely to be accepted?

glasser commented:
I've noticed a bit of a bug in here.  Let's say you deploy pointing to one image.  Then you change the image in the TF source file and also make an out-of-band change such that the old image no longer exists. This will lead to an error at refresh time.

I think this means that for providers like this one, the Read step should be a no-op... as long as you can be confident that once it succeeds it will always succeed.

Ideally what I'd want is a way to make it so that refresh/Read is a no-op if the resource is going to be replaced anyway, but actually does the read if it is not going to be replaced. Is that possible?

glasser commented:
Changed so that Read is a no-op.

glasser commented:
I've updated this PR to create a data source rather than a resource, which is more appropriate.  I've updated the vendoring to use govendor rather than godep.  It is still missing docs, which I'm happy to provide if this is likely to be accepted.

Note that it now uses the V2 registry API rather than the V1 API.

glasser commented:
Resolved the trivial merge conflict.

stack72 commented:
Hi @glasser 

Now that this is a data_source, could you rebase this and we can get it tested and (potentially) merged?

P.

paddycarver commented:
Hi! This issue has been open for some time and waiting response from the community without any activity, so I'm going to close it out. We do this to try to keep our attention on PRs that are actively in-progress, it's not a comment on the code. If there is still interest in this PR, please feel free to reply, and we'll re-open this. Thanks!

glasser commented:
I totally get that! I've kept this issue at the top of my non-work-urgent to-do list for a long time now and plan to resurrect this once my company upgrades to TF 0.9.

paddycarver commented:
Awesome! Thanks for the interest in contributing, @glasser. Totally understand that time is hard to come by. Whenever you're ready, do you just want to ping on this PR, and we'll get it reopened? Thanks!
